src/components/settings/PushNotifications/logic.ts

```diff
@@ -0,0 +1,118 @@
+import { keccak256, toUtf8Bytes } from 'ethers/lib/utils'
+import { getToken, getMessaging } from 'firebase/messaging'
+import { DeviceType } from '@safe-global/safe-gateway-typescript-sdk/dist/types/notifications'
+import type { RegisterNotificationsRequest } from '@safe-global/safe-gateway-typescript-sdk/dist/types/notifications'
+import type { Web3Provider } from '@ethersproject/providers'
+
+import { FIREBASE_MESSAGING_SW_PATH, FIREBASE_VAPID_KEY } from '@/config/constants'
+import { trackEvent } from '@/services/analytics'
+import { PUSH_NOTIFICATION_EVENTS } from '@/services/analytics/events/push-notifications'
+import packageJson from '../../../../package.json'
+
+type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] }
+
+// We store UUID locally to track device registration
+export type NotificationRegistration = WithRequired<RegisterNotificationsRequest, 'uuid'>
+
+export const requestNotificationPermission = async (): Promise<boolean> => {
+  if (Notification.permission === 'granted') {
+    return true
+  }
+
+  let permission: NotificationPermission | undefined
+
+  try {
+    permission = await Notification.requestPermission()
+  } catch (e) {
+    console.error('Error requesting notification permission', e)
+  }
+
+  const isGranted = permission === 'granted'
+
+  trackEvent(isGranted ? PUSH_NOTIFICATION_EVENTS.GRANT_PERMISSION : PUSH_NOTIFICATION_EVENTS.REJECT_PERMISSION)
+
+  if (!isGranted) {
+    alert('You must allow notifications to register your device.')
+  }
+
+  return isGranted
+}
+
+const getSafeRegistrationSignature = ({
+  safeAddresses,
```

Comment:
Is there a case where we have more than one Safe address for the registration that requires a signature?

Reply:
No. In the current design, you can only sign for notifications within a Safe.